### ISSUE TITLE
Alternative training labels

### ISSUE THREAD

**Issue by angusolav** [DeepMReye/DeepMReye#85]
Hi! As FAQ page is down, I hope it is ok that I ask a question here. I want to train my own model but dont have 10 calibration coordinates with fMRI scans, we unfortunately only started capturing fMRI after the calibration procedure with the Eyelink 1000. However I do have the eye tracking from the Eyelink during task performance in the MRI. So is it possible for me to for example identify some participants with the best eyetracker calibration, identify 10 coordinates that they all looked at during task that cover the screen they were looking and use these coordinates + the fMRI captured at the same time as alternative training labels? Or will the fact that the eye tracking data I get after calibration is less reliable make model training with this data much worse? Any thoughts on the feasibility of this? Thnaks.

**Comment by Naubody** [comment]
This should work! Though you may not even need to limit the training labels to 10 positions, but instead try training it on the full eye tracking data of all participants. Just clean these data thoroughly as you do not want tracking artefacts etc. in the training labels. If in doubt, exclude more data rather than less. 

How many participants do you have, and how much eye-tracking data per participant? What kind of viewing behavior is it (guided fixations, free viewing, smooth pursuit...)?

**Comment by angusolav** [comment]
Great! I think I have about 30 participants now with eye tracker but more where the eye tracker didnt work well, which is why I want to use the deepmreye model. I only have 15 participants with good calibration (avg. error < 0.5 and maximum error < 1 degree of visual angle).  But I will be getting more participants in the coming months. For info, we used the Eye-link 1000 at 1000 hertz sampling rate concurrently with 7T MRI data aquisition. The task was rating facial expressions (happy, fearful, neutral). They would first be presented a cross in the middle of the screen for 6 seconds then presented a face for 1 second.  They would subsequently be presented a prompt for 3 seconds asking for response on whether the face was happy/fearful/neutral and then another prompt for 3 seconds asking for confidence 1/2/3. The obvious problem here is a lack in diversity in gaze across the screen. During fixation on the cross and responding to the prompts, I assume the gaze is primarily in the middle of the scree. Only during the 1 seconds showing a face should there be real diversity in where they gaze, but again I think this might be limited. Luckily however we did subject most of these participants to movie watching as well. Although this is not the data I was intending to analyze, maybe it is more practical to use this for model training and it will still be generalizable to their viewing during the face task? Then I would have a lot of training data, about 20 minutes of eye tracking during movie watching. And just to be sure, I do not need to shift the eye tracking data to correspond with heomodynamic response right? The model uses instantaneous structural changes of the eyes, not the blood oxygenation?

**Comment by Naubody** [comment]
Some development here over email.

The question was the following: 
"The EyeLink 1000 has captured gaze position concurrently at a 1000 hertz sampling rate. My issue is that there is great variation in gaze position within a TR. In other words, how can I assign specific gaze positions to a functional volume when the functional volume is collected over 2 seconds and thus contains large variations in gaze positions?"

Answer: 
By default, DeepMReye decodes, and is trained on, 10 gaze labels per functional volume. For your repetition time of 2s, this effectively means that you need to downsample your 2000 Hz eye-tracking time series to 5 Hz, and then chunk it into blocks of ten labels each corresponding to your functional volumes. Now that 5 Hz may sound low, but remember that humans only perform around 5 fixations per second anyways. This means that a 5 Hz eye tracker can capture the large majority of the variance in your gaze path, with higher resolutions quickly plateauing in how much more variance is explained. Unless you study the precise curvature of saccades or the alike, your 2000 Hz eye tracking will not be necessary. Further note that even for free viewing of movies, even a single gaze position per functional volume explains a lot of variance in the eye tracking data (certainly >80% in most datasets I have seen), so even averaging the 10 labels again after decoding can still enable many useful analyses.

**Comment by Naubody** [comment]
Added this and another related question to our new [FAQ page](https://github.com/DeepMReye/DeepMReye/wiki/DeepMReye-%E2%80%90-FAQ).

**Comment by angusolav** [comment]
Thanks for the explanation! I havent had the time to try this out until now so sorry for the late reply. I am still a little confused how to do this. You say that the model can be trained on 10 gaze labels per functional volume: TR = 2s will require 5hz eye-tracking data, 10 gaze coordinates per volume. What I dont understand is how exactly to go about this as it seems that your model requires 1 gaze label per functional volume. It seems like your code, for example in the save_data() function, expect the gaze label to be in shape (t,2) where t corresponds to each functional volume, therefore 1 xy coordinate per volume. My understanding is that if I want to use 10 labels per volume, the shape of the gaze labels must then be (t, 10, 2) which the DeepMReye code cant handle? What am I missing? Am I rather supposed to reiterate matching between a functional volume and a gaze label 10 times so that each volume is duplicated 10 times with the 10 different gaze coordinates? Appreciate your help! Thanks

**Comment by CYHSM** [comment]
Hey @angusolav, thanks for your question. As you mentioned you need 10 gaze labels per TR, the sampling rate of the eye tracking data is not that important as we evenly sample (N=_inner_timesteps_) from them, see here: https://github.com/DeepMReye/DeepMReye/blob/9318b4f22f5c85b408036d48d5db6963b729960a/deepmreye/util/data_generator.py#L248-L249
the inner_timesteps options defines the number of gaze labels which the model is trained on. The model expects the gaze labels to be in the shape (t, inner_timesteps, 2), so if you only have one gaze label it would be (t,1,2) and for most of the results in the paper we use (t, 10, 2). 

Regarding your question on how this works, there is no need to duplicate the volume. What is duplicated automatically are the regression blocks in the model (it dynamically adapts to inner_timesteps, see below) and then just predicts as many gaze coordinates as you have for this TR.  https://github.com/DeepMReye/DeepMReye/blob/9318b4f22f5c85b408036d48d5db6963b729960a/deepmreye/architecture.py#L141-L157 

Regarding the save_data function, this should work fine with (t, 10, 2) as we use it here: https://github.com/DeepMReye/DeepMReye/blob/9318b4f22f5c85b408036d48d5db6963b729960a/streamlit/streamlit.py#L114  Let me know if this does not work for you then we might need to adapt it!

**Comment by CYHSM** [comment]
Closing due to inactivity, feel free to repoen @angusolav or create a new issue, thanks.
